src/metabase/models/query/permissions.clj

```diff
@@ -59,21 +60,41 @@
 ;;           {:perms/view-data {table-id :unrestricted}}  source-card-read-perms
 ;;
 
-(mu/defn query->source-table-ids :- [:set [:or [:= ::native] ::lib.schema.id/table]]
-  "Return a sequence of all Table IDs referenced by `query`, and/or the ::native keyword for native queries."
+(defn- merge-source-ids
+  "Merge function which takes the union of two sets of IDs, if they are both sets"
+  [val1 val2]
+  (if (and (set? val1) (set? val2))
+    (set/union val1 val2)
+    val2))
+
+(mu/defn query->source-ids :- [:maybe
+                               [:map
+                                [:table-ids {:optional true} [:set ::lib.schema.id/table]]
+                                [:card-ids  {:optional true} [:set ::lib.schema.id/card]]
+                                [:native?   {:optional true} :boolean]]]
+  "Return a map containing of table IDs and/or card IDs referenced by `query`, and/or the :native? boolean flag
+  indicating a native query or subquery. Intended to be used in the context of permissions enforcement."
   [query :- :map]
-  (set
-   (flatten
-    (lib.util.match/match query
-      ;; if we come across a native query just put a placeholder (`::native`) there so we know we need to
-      ;; add native permissions to the complete set below.
-      (m :guard (every-pred map? :native))
-      [::native]
+  (apply merge-with merge-source-ids
+         (lib.util.match/match query
+           ;; If we come across a native query, replace it with a card ID if it came from a source card, so we can check
+           ;; permissions on the card and not necessarily require full native query access to the DB
+           (m :guard (every-pred map? :native))
+           (if-let [source-card-id (:qp/stage-is-from-source-card m)]
```

Comment:
Ah, I guess this answers my question from above. We only collect `:card-ids` from `:native` stages.

Is there any case where `:qp/stage-is-from-source-card` can appear on a non-native stage and, if so, would we want to collect those as well? Like would it make sense to turn this `if-let` into a `when-not`, then have a separate match guard that matches `:qp/stage-is-from-source-card` anywhere in the query?

Oh, maybe not because if the expanded card query isn't native, you'll have a `:source-table` to collect instead?

Reply:
Confirmed in [this thread](https://metaboat.slack.com/archives/C064EB1UE5P/p1730998600149119). Updating the PR shortly